### ISSUE TITLE
some requests

### ISSUE THREAD

**Issue by greyltc** [uio-ess/data-coolection#1]
1. please add a `photons_per_count` float attribute in the camera section, use `None` if it's not known and my analysis will compute everything in camera counts
1. please add a human readable `session` string attribute at the root level. for now, i've hardcoded my anaysis to use `FEB '18 OCL`. trigger_id should not be reset during a session, so the session string and trigger_id are enough to uniquely identify every data collection event.

**Comment by hgjersdal** [comment]
Not sure about 1, is it not better to store this value some other place that is easier to update? The setup  we used for the calibration has room for improvement, it is a value that is likely to change.
What we for sure should add to the camera section is mandatory fields for lens name, focal length and f-number. I will add this together with your other suggestions to the experimental branch later today.

**Comment by greyltc** [comment]
It's easy to update!

**Comment by hgjersdal** [comment]
It is easy to update a field in a file, but we would need to update a bunch of files on a bunch of computers. For example lens transmission varies with time, so a calibration would need to be matches with a specific camera and a time window, so maybe a list of sessions. If we want to be really accurate, we may need to match it with wavelength too, the quantum efficiency is not linear. Are we sure this is the only calibration we need? Might be nice with a pointer to relevant dark frames, for example.  Maybe pixel masks. 

Would it not be easier to just just read it from something like FEB-18-OCL-calib.h5? We could put it on git, all you would need to do is pull in order to be sure you have the latest calib.

**Comment by greyltc** [comment]
Using git for binary files is generally a bad idea. I really dislike the idea of looking anywhere but inside the run file during analysis, especially for the kind of real-time analysis I'm trying to work towards. I'm very much trying to make these files 100% stand-alone.

There's only one place the files need to be updated and that's the master copy on rsync.

**Comment by hgjersdal** [comment]
Fine, you're running the analysis show, I'll add it. If we are planning on people from several places rewriting calib info to the files multiple times, we for sure should have a backup of the raw, untuoched hdf5 files somewhere. I'll check with Ole how he is doing on the raid. 

What about dark frames, pixel masks etc, more accurate photons per count calibration and so on, though? Adding those to every file sounds silly. You may not want to use that on the file by file summaries, but for publications we are going to need proper calibration. We should have some session calibration hdf5 files, right? Probably on rsync, not git.

**Comment by greyltc** [comment]
I can imagine my analysis routine looking into the H5 file at the time of processing for a `/calibration` group. If that group exists, the calibration is applied. If not, then the analysis is run uncalibrated. If we wish to have analysis that takes the calibration data into account, then we feed it files which contain calibration data. We could insert the cal data into the data files as needed.

We should probably talk about a few of the features provided by rsync.net
Including calibrationin every single file *should* incur almost no storage penalty in our rsync account. rsync.net is built on the ZFS file system which has a deduplication mechanism that means redundant data is stored only once. (I'll check this assumption I know it works this way on my server at home). So I do advocate putting the cal data into every single file.
Also, there are 7 daily "snapshots" over all the data in our account. Anything we mess up can be recovered by reading back one of these snapshots. If 7 days goes by without us realizing we've made a mistake, the change can not be undone or if we mess the data up before it's been snapshotted we're out of luck. Let's not mess the data up and let's be careful not to rewrite the files simultaneously. We could have a lock file that prevents us from being dumb. So *maybe* additional backup is not needed.

I'd eventually like to have a sanitize routine that runs on all the as-collected h5 files, and checks if they're bad. I get a bunch of bad files in my analysis now (from things like ctrl-c while writing the files I guess) Handling all of the cases where the files can break is making my analysis code a bit convoluted right now. I'd like to break that logic out into this sanitize or data "washing" routine, having that also insert calibration data seems pretty good to me. This routine could even run in the background on the data collection machine and wash the data in real-time before it even goes up to rsync.net.

Let's not do anything that prevents us from doing publication quality analysis in real-time. If good cal data is not ready at the time of data collection, then fine, but if it is, let's use it!

**Comment by hgjersdal** [comment]
Ok, if copying the calib data to all the files does not take up rsync space, then copying it is fine. But having an attribute with a calibration file name, and looking for that file in your synced calibration file folder also seems pretty straight forward. It could also save space on the laptops where the data is synced to.

If it would help, I can add an attribute called 'write_finished', and set it to false when I open the file, and true just before closing it. Should make it easy to spot C-c files. (BTW writing None to attributes does not work, False does.)

If we get the files on a raid we can also potentially save space on rsync by deleting files marked as bad and still have them. I've already done dumb things, some of them subtle, to hdf5 files during generation, not sure I can guarantee avoid it during updating. It is nice to have the security of a backup of the original files.

**Comment by greyltc** [comment]
Yeah, having some server somewhere do a one-way rsync download every 5min of everything in pur rsync.net wouldn't hurt at all.

I like the idea of a `finished` attribute that is false when you open the file and set to true just before you're done.

**Comment by hgjersdal** [comment]
Camera now has attribute photons_per_count, initialized by __init__ argument,  set to False by default.

All devices now have 'Capture time' attribute. Set from callbacks where applicable, or call to time.time() during write for things like the stage and the temperature controller.

In the experimental branch.
